master/buildbot/changes/gerritchangesource.py

```diff
@@ -154,22 +186,49 @@ def getGroupingPolicyFromEvent(self, event):
                               event_change['number'])
         return event_change["branch"]
 
+    @defer.inlineCallbacks
     def addChangeFromEvent(self, properties, event):
+        if "change" not in event:
+            if self.debug:
+                log.msg("unsupported event %s" % (event["type"],))
+            return defer.returnValue(None)
+
+        if "patchSet" not in event:
+            if self.debug:
+                log.msg("unsupported event %s" % (event["type"],))
+            return defer.returnValue(None)
 
-        if "change" in event and "patchSet" in event:
-            event_change = event["change"]
-            return self.addChange({
-                'author': _gerrit_user_to_author(event_change["owner"]),
-                'project': util.bytes2unicode(event_change["project"]),
-                'repository': "{}/{}".format(
-                    self.gitBaseURL, event_change["project"]),
-                'branch': self.getGroupingPolicyFromEvent(event),
-                'revision': event["patchSet"]["revision"],
-                'revlink': event_change["url"],
-                'comments': event_change["subject"],
-                'files': ["unknown"],
-                'category': event["type"],
-                'properties': properties})
+        event = canonicalize_event(event)
+        event_hash = hash_event(event).hexdigest()
+        is_new_event = yield (self.master.db.gerriteventhashes
+                              .insertEventHash(event_hash))
+
+        event_change = event["change"]
+        if not is_new_event:
+            if self.debug:
+                eventstr = "{}/{} -- {}:{}".format(
+                    self.gitBaseURL, event_change["project"],
+                    self.getGroupingPolicyFromEvent(event),
+                    event["patchSet"]["revision"])
+                message = (
+                    "gerrit: duplicate change event {} by {}"
+                    .format(eventstr, self.__class__.__name__))
+                log.msg(message.encode("utf-8"))
+            return defer.returnValue(None)
+
+        cb = yield self.addChange({
+            'author': _gerrit_user_to_author(event_change["owner"]),
```

Comment:
nitpick: some consistency about quote usage would be appricated

Reply:
Done